framework/src/meshgenerators/BlockToMeshConverterGenerator.C

```diff
@@ -0,0 +1,125 @@
+      "ID)");
+
+  return params;
+}
+
+BlockToMeshConverterGenerator::BlockToMeshConverterGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    _input(getMesh("input")),
+    _target_blocks(getParam<std::vector<SubdomainName>>("target_blocks"))
+{
+}
+
+std::unique_ptr<MeshBase>
+BlockToMeshConverterGenerator::generate()
+{
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+
+  if (!mesh->is_replicated())
+    mooseError("BlockToMeshConverterGenerator is not implemented for distributed meshes.");
+
+  auto new_mesh = buildMeshBaseObject();
+
+  std::vector<subdomain_id_type> target_block_ids =
+      MooseMeshUtils::getSubdomainIDs((*mesh), _target_blocks);
+
+  // Check that the block ids/names exist in the mesh
+  std::set<SubdomainID> mesh_blocks;
+  mesh->subdomain_ids(mesh_blocks);
+
+  for (std::size_t i = 0; i < target_block_ids.size(); ++i)
+    if (target_block_ids[i] == Moose::INVALID_BLOCK_ID || !mesh_blocks.count(target_block_ids[i]))
+    {
+      paramError("target_blocks",
+                 "The target_block '",
+                 getParam<std::vector<SubdomainName>>("target_blocks")[i],
+                 "' was not found within the mesh.");
+    }
+
+  // know which nodes have already been inserted, by tracking the old mesh's node's ids'
+  std::unordered_map<dof_id_type, dof_id_type> old_new_node_map;
+
+  for (subdomain_id_type target_block_id : target_block_ids)
+  {
+
+    for (auto elem : mesh->active_subdomain_elements_ptr_range(target_block_id))
+    {
+      // make a deep copy so that mutiple meshs' destructors don't segfault at program termination
+      auto copy = elem->build(elem->type());
+
+      // index of node in the copy element must be managed manually as there is no intelligent
+      // insert method
+      dof_id_type copy_n_index = 0;
+
+      // correctly assign new copies of nodes, loop over nodes
+      for (dof_id_type i : elem->node_index_range())
+      {
+        auto & n = *elem->node_ptr(i);
+
+        if (old_new_node_map.count(n.id()))
+        {
+          // case where we have already inserted this particular point before
+          // then we need to find the already-inserted one and hook it up right
+          // to it's respective element
+          copy->set_node(copy_n_index++) = new_mesh->node_ptr(old_new_node_map[n.id()]);
+        }
+        else
+        {
+          // case where we've NEVER inserted this particular point before
+          // add them both to the element and the mesh
+
+          // Nodes' IDs are their indexes in the nodes' respective mesh
+          // If we set them as invalid they are automatically re-assigned
+          // to be the last element in the mesh's node array
+          std::unique_ptr<Node> node_copy = Node::build(elem->point(i), Node::invalid_id);
+
+          // Add to mesh. This method call will automatically re-assign the id; so must call this
+          // first
+          new_mesh->add_node(node_copy.get());
```

Comment:
This is kind of dangerous, leaving two `unique_ptr` pointing to the same target and then manually releasing the latter.  It's also kind of weird, manually creating the Node yourself without setting an id.  Why not just `Node * node_copy = new_mesh->add_point(elem->point(i));`?

Reply:
>This is kind of dangerous, leaving two unique_ptr pointing to the same target and then manually releasing the latter. 

The `.get()` method returns, not another unique_ptr -- but a raw pointer of what the unique_ptr is owning. 

> It's also kind of weird, manually creating the Node yourself without setting an id. 

I thought that the `add_node` method would automatically assign the id, correctly. When I tried to assign it myself it didn't work for me. From what I gathered, the id gets used as the index of the node in the list kept internally by the mesh. Therefore the id must be unique, and ordered correctly. The `i` variable loops through the list of nodes in each element, even while the elements were being looped over. There was no way for me to choose what order the elements were looped over, to ensure that the point I was `add_point`-ing had an id that was in ascending order in the new mesh. 

Hope that made some kind of sense. There could be something I'm not understanding here. Is it important to keep the node id's consistent from one mesh to the next?

> Why not just `Node * node_copy = new_mesh->add_point(elem->point(i));`?

That is an elegant alternative. I'm way pre-maturely optimizing here, but I speculated that just re-assigning from the ptr would be more efficient, since it wouldn't require creating a new variable, which would take up another register in the CPU. (probably silly, but that was my rationale)

Reply:
> The .get() method returns, not another unique_ptr -- but a raw pointer of what the unique_ptr is owning.

Sure.  And when you pass that raw pointer to `add_node`, you're promising us that you allocated it (true!) and that it's our responsibility to delete it (false! total lie! you've still got another deleter assigned to it!).  Our fault, for not documenting that well enough, but it's still a code smell.  You manually release the redundant deleter afterward, but the whole point of smart pointers is to make it so those dangerous situations provably never happen.

> I thought that the add_node method would automatically assign the id, correctly.

"Correctly" is a matter of context, but in the context here it'll do fine, giving you an id one past the maximum existing id.  My point is that that's *also* what `add_point` would do, with less work in your code.  You just get the `Node *` returned by `add_point` and check that to see what `id()` it got assigned.  You've got 4 lines of code here that could be replaced by 3; that's not a problem because that 4th line is going to take up too much hard drive space or anything, but just because it's misleading, it suggests that something special is being done here.

... likewise for `copy_n_index`, actually, now that I look at that.  Isn't that always going to be the same as `i`?

> I'm way pre-maturely optimizing here

Yeah.  `add_node` and `add_point` are both virtual function calls (they're inserting into different data structures depending on whether your mesh is `ReplicatedMesh` or `DistributedMesh`); even if your code did less work that would be dwarfed by the virtual thunk and stack manipulation.